File: game/src/main/scala/io/aigar/game/Game.scala
package io.aigar.game

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.blocking
import scala.math.round
import com.github.jpbetz.subspace.Vector2
import com.typesafe.scalalogging.LazyLogging
import io.aigar.score.ScoreModification
import io.aigar.controller.response.Action

/**
 * Game holds the logic for an individual game being played
 * (e.g. the ranked game or a private test game).
 */
object Game {
  final val RankedGameId = -1
  final val DefaultDuration = 60 * 20
  final val DefaultMutliplier = 1
  final val PrivateGameDuration = 60 * 10
  final val PrivateGameBotQuantity = 5
  final val MinimumNumberOfPlayerModificator = 10

  final val NanoSecondsPerMillisecond = 1000000f
  final val MillisecondsPerSecond = 1000f
  final val NanoSecondsPerSecond = NanoSecondsPerMillisecond * MillisecondsPerSecond

  final val TicksPerSecond = 15
  final val MillisecondsPerTick = round(MillisecondsPerSecond / TicksPerSecond)

  /**
    * Current time, in seconds.
    */
  def time: Float = {
    System.nanoTime / NanoSecondsPerSecond
  }
}

class Game(val id: Int,
           playerIds: List[Int],
           val duration: Int = Game.DefaultDuration,
           val multiplier: Int = Game.DefaultMutliplier)
    extends LazyLogging {
  logger.info(s"Launching game with ID $id and $multiplier multiplier.")

  val grid = new Grid(math.max(Game.MinimumNumberOfPlayerModificator, playerIds.length) * Grid.WidthPerPlayer,
                      math.max(Game.MinimumNumberOfPlayerModificator, playerIds.length) * Grid.HeightPerPlayer)
  val players = createPlayers
  val viruses = new Viruses(grid, math.max(Game.MinimumNumberOfPlayerModificator, playerIds.length))
  val resources = new Resources(grid)

  val startTime = Game.time
  var previousTime = startTime
  var currentTime = startTime + Game.MillisecondsPerTick / Game.MillisecondsPerSecond // avoid having an initial 0 delta time
  var pausedTime = 0f
  var tick = 0

  def update: Future[(List[ScoreModification], serializable.GameState)] = {
    Future {
      this.synchronized {
        blocking {
          val deltaSeconds = currentTime - previousTime
          var modifications = players.flatten {  _.update(deltaSeconds, grid, players) }
          modifications :::= viruses.update(grid, players)
          modifications :::= resources.update(grid, players)
          tick += 1

          previousTime = currentTime
          currentTime = Game.time
          (modifications, state)
        }
      }
    }
  }

  def updatePaused: Unit = {
    pausedTime += currentTime - previousTime

    previousTime = currentTime
    currentTime = Game.time
  }

  def performAction(player_id: Int, actions: List[Action]): Future[List[ScoreModification]] = {
    Future {
      this.synchronized {
        blocking {
          players.find(_.id == player_id) match {
            case Some(player) => {
              player.performAction(actions)
            }
            case None => List()
          }
        }
      }
    }
  }

  def timeLeft: Float = {
    duration + pausedTime - (Game.time - startTime)
  }

  def state: serializable.GameState = {
    serializable.GameState(
        id,
        false,
        multiplier,
        tick,
        timeLeft,
        players.map(_.state),
        resources.state,
        grid.state,
        viruses.state
      )
  }

  def createPlayers: List[Player] = {
    playerIds.map { new Player(_, spawnPosition) }
  }

  def spawnPosition: Vector2 = {
    grid.randomPosition
  }
}

File: clients/javascript/game/game.js
const PlayerError = require('./errors.js').UnknownPlayerIdException;
const Player = require('./player.js');
const Map = require('./map.js');
const Resources = require('./resources.js');
const Virus = require('./virus.js');

class Game {
  constructor(playerId, id, tick, timeLeft, players, resources, map, viruses) {
    this.id = id;
    this.tick = tick;
    this.timeLeft = timeLeft;
    this.players = players;
    this.resources = resources;
    this.map = map;
    this.viruses = viruses;

    this.me = players.find(p => p.id === playerId);
    if(this.me === undefined) {
      throw new PlayerError();
    }

    this.enemies = players.filter(p => p.id !== playerId);
  }

  static parse(payload, playerId) {
    return new Game(
      playerId,
      payload.id,
      payload.tick,
      payload.timeLeft,
      payload.players.map(p => Player.parse(p)),
      Resources.parse(payload.resources),
      Map.parse(payload.map),
      payload.viruses.map(v => Virus.parse(v))
    );
  }

  static get RANKED_GAME_ID() {
    return -1;
  }

  static get UPDATE_PER_SECOND() {
    return 3;
  }

  get actions() {
    return this.me.actions;
  }
}

module.exports = Game;
